### PR TITLE
[ROCm] add hipblaslt support

### PR DESCRIPTION
**Pull request by jeffdaily**
Disabled by default. Enable with env var DISABLE_ADDMM_HIP_LT=0. Tested on both ROCm 5.7 and 6.0.

cc @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

### CODE REVIEW

File: aten/src/ATen/cuda/CUDABlas.cpp [hunk 1]
```diff
@@ -398,7 +409,13 @@ void bgemm<at::BFloat16>(CUDABLAS_BGEMM_ARGTYPES(at::BFloat16)) {
                                   (void*)&falpha, a, CUDA_R_16BF, (int)lda, stridea,
                                   b, CUDA_R_16BF, (int)ldb, strideb,
                                   (void*)&fbeta, c, CUDA_R_16BF, (int)ldc, stridec,
-                                  (int)num_batches, CUDA_R_32F, CUBLAS_GEMM_DEFAULT_TENSOR_OP));
+                                  (int)num_batches,
+#if defined(USE_ROCM) && ROCM_VERSION >= 60000
+                                  CUBLAS_COMPUTE_32F,
```
Comment (xw285cornell) [inline aten/src/ATen/cuda/CUDABlas.cpp:414]: Actually I get an error here: 
no known conversion from 'hipblasLtComputeType_t' to 'hipblasDatatype_t' for 22nd argument

Even if we use the v2 api of hipblasGemmStridedBatchedEx_v2, it takes hipblasComputeType_t on my 6.0.0 header file.

File: aten/src/ATen/cuda/CUDABlas.cpp [hunk 2]
```diff
@@ -588,12 +605,22 @@ void gemm<at::BFloat16>(CUDABLAS_GEMM_ARGTYPES(at::BFloat16)) {
       c,
       CUDA_R_16BF,
       ldc,
+#if defined(USE_ROCM) && ROCM_VERSION >= 60000
+      CUBLAS_COMPUTE_32F,
```
Comment (xw285cornell) [inline aten/src/ATen/cuda/CUDABlas.cpp:609]: this is still hipblas api, are we supposed to use hipblaslt compute type here?

**Review by xw285cornell — COMMENTED** [review]
Another common issue I'm seeing is "assigning to 'hipDataType' from incompatible type 'hipblasDatatype_t'"

File: aten/src/ATen/cuda/CUDABlas.cpp (hunk 1 above)
Comment (jeffdaily) [reply aten/src/ATen/cuda/CUDABlas.cpp:414]: There is a hipblaslt API mismatch happening.  This PR was verified against stock ROCm 5.7, as well as a build of ROCm 6.0 that is between RC2 and RC4 -- in anticipation of what the final ROCm 6.0 release should look like.  The tip of tree develop branch of hipblaslt is not yet aligned to our intended API that will be part of 6.0.  If you were to build this hipblaslt PR, it represents what our 6.0 release should look like (and will work with this PR).

https://github.com/ROCmSoftwarePlatform/hipBLASLt/pull/430

File: aten/src/ATen/cuda/CUDABlas.cpp [hunk 3]
```diff
@@ -176,18 +180,25 @@ uint32_t _getAlignment(uintptr_t address) {
...
   }
 }
+#endif
 
 static size_t _parseChosenWorkspaceSize() {
   const char * val = getenv("CUBLASLT_WORKSPACE_SIZE");
+#ifdef USE_ROCM
+  if (!val) {
+    // accept either env var
+    val = getenv("HIPBLASLT_WORKSPACE_SIZE");
```
Comment (jianyuh) [inline aten/src/ATen/cuda/CUDABlas.cpp:190]: What's the recommended hipsblaslt workspace size for the best perf ?

File: aten/src/ATen/cuda/CUDABlas.cpp (hunk 1 above)
Comment (jianyuh) [reply aten/src/ATen/cuda/CUDABlas.cpp:414]: Just conforming, this PR relies on the latest commit of hipBLASLt (e.g., https://github.com/ROCmSoftwarePlatform/hipBLASLt/pull/430 is committed 5 days ago)? The current public release version of hipBLASLt doesn't work.

File: aten/src/ATen/cuda/CUDABlas.cpp (hunk 2 above)
Comment (jianyuh) [reply aten/src/ATen/cuda/CUDABlas.cpp:609]: In cuBLAS vs. cuBLASLt, I see the data types for compute type are the same:

https://docs.nvidia.com/cuda/cublas/#cublascomputetype-t

Ideally hipBLASLt should just follow this convension.

**Comment by jianyuh** [comment]
@jeffdaily Do you have any benchmark evaluation results on torch.mm perf ?

**Comment by xw285cornell** [comment]
Some new set of failures: no known conversion from 'const hipblasDoubleComplex *' to 'const hipDoubleComplex *

File: aten/src/ATen/cuda/CUDABlas.cpp (hunk 2 above)
Comment (jeffdaily) [reply aten/src/ATen/cuda/CUDABlas.cpp:609]: @jianyuh in ROCm 6.0, the hipblaslt APIs are updated to align with cuBLAS/cuBLASLt w.r.t. types.  In the meantime, this PR tries to strike a balance between the state of the hipblaslt API in ROCm 5.7, hipblaslt develop branch, and hipblaslt 6.0 branch.  It is painful that the APIs changed so much, but if we can live with the workarounds in this PR for now that would be helpful.

File: aten/src/ATen/cuda/CUDABlas.cpp (hunk 1 above)
Comment (jeffdaily) [reply aten/src/ATen/cuda/CUDABlas.cpp:414]: This PR should not rely on hipblaslt PR 430.  It should work with hipblaslt from its 5.7 release branch, 6.0 release branch, and develop branch.  Unfortunately, all 3 hipblaslt branches have API differences so this PR works around this.

**Review by malfet — APPROVED** [review]

**Comment by atalman** [comment]
@jeffdaily please attach ciflow/trunk label when trying to reland

**Comment by jeffdaily** [comment]
@atalman are you sure it was this PR that broke the HUD?  The HUD link for the trunk merge commit doesn't show any failures.

https://hud.pytorch.org/pytorch/pytorch/commit/bb2bb8cca1c00e3f6e7025a62688d0cfcbfee144

**Comment by jeffdaily** [comment]
@atalman ok I see it now.  The windows build failed.

https://ossci-raw-job-status.s3.amazonaws.com/log/19659846551

**Comment by xw285cornell** [comment]
fyi this PR will break internal test so I'll need to make some changes in our code. Let me know when you're going to re-land this.

**Comment by xw285cornell** [comment]
@jeffdaily sorry it seems I break some tooling. I'll do another land

**Pull request closed without merging**